### DISCUSSION TITLE
Limited Azure UDP MTU prevents successful streaming from Azure devices

### DISCUSSION THREAD

**Discussion by neilyoung** [bluenviron/mediamtx#1588] [General]
Have you ever heard about this? 

For now it is just a symptom: I'm having identical installations using your RTSP server on Azure and AWS. Only on AWS I'm able to have UDP clients. In local traces I see, that packets form the RTSP server come in as "Fragmented IP" on AWS, nothing like this on Azure. There only scattered packages come in, most of the clients complain about tons of lost packages.

I suspect the problem in the 1472 payload, which seems to be too big for Azure. 

TLS works on both fine, so I can definitely exclude the source, which feeds the RTSP server. It is the client side end, which makes trouble on Azure.

I see two places, where this packet size is hardcoded:

https://github.com/aler9/rtsp-simple-server/blob/8ad376de53dc0c0741c357770ebd3bd0ae7d5c3e/internal/formatprocessor/generic.go#L13

https://github.com/aler9/rtsp-simple-server/blob/fb1c16d60627bfa6d34814077542993b9c98af2c/internal/core/udp_source.go#L130

Any chance to have that configured from a variable for a test?

**Comment by neilyoung** [comment]
What I can see (tcpdump on the Azure and AWS server outgoing interface, RTP UDP stream requested by client):

Azure: No fragmentation

<img width="1630" alt="image" src="https://user-images.githubusercontent.com/731020/227036732-250c7868-24e3-4807-94e7-eed245f32fbf.png">

"DF" bit set in IP header
<img width="609" alt="image" src="https://user-images.githubusercontent.com/731020/227036884-229311f4-e053-4312-afdb-b635e1f9e972.png">

AWS: Fragmentation

<img width="1799" alt="image" src="https://user-images.githubusercontent.com/731020/227039396-db30e87b-7310-440f-940f-0e45a0d2a9df.png">

"DF" bit not set in IP header

<img width="558" alt="image" src="https://user-images.githubusercontent.com/731020/227039518-64915150-3a03-4c19-b72e-652503ac24ca.png">

Azure UDP packets are missing at the client, AWS are not. Does that ring a bell?

**Comment by neilyoung** [comment]
Nobody having a similar issue with Azure and RTSP UDP?

**Comment by aler9** [comment]
Hello, it's clear that Azure limits the UDP MTU to a lower value than its usual one, and this causes problems, as always UDP when fragmentation is involved. We can add an option inside the server in order to decrease the MTU.

The only issue is, we can certainly tune the MTU of UDP packets sent to clients that are reading, but i don't know how to instruct clients that are publishing with RTSP/UDP to decrease the MTU of their outgoing packets. Will investigate further.

**Comment by neilyoung** [comment]
> We can add an option inside the server in order to decrease the MTU.

How would you do that? Just curious. You talk about the MTU or the packet size (1472)? A means to play with the packet size by configuration would be awesome. Would offer test support.

What I don't understand is, they say, they have an MTU of 1500 and fragment that to 1400. But I can't see this. The packet sent is 1514 in total (even higher as AWS with 1506), but no fragmentation is attempted.

RTP packets from client to RTSP server is currently not my construction site, but I remember to have seen packet size options for GStreamer...

In fact, this flaw (?) currently effectively prevents UDP streaming from Azure to any client. TCP via TLS works of course.

**Reply by aler9** [reply]
> . The packet sent is 1514 in total (even higher as AWS with 1506), but no fragmentation is attempted.

They're probably adding padding or another layer, i can't say anything since i didn't see the entire packet dump.

> How would you do that?

The server is already able to fragment packets bigger than 1472 bytes into smaller packets:
https://github.com/aler9/rtsp-simple-server/blob/5ffe8b2fc47e86e08799c0dba227335ec1ad280a/internal/formatprocessor/h264.go#L214

therefore it's enough to link maxPacketSize to the configuration.

**Reply by neilyoung** [reply]
> They're probably adding padding or another layer, i can't say anything since i didn't see the entire packet dump.

Would that help if I would provide you one? I could take it on the Azure output side for instance or on the input of my device. Just RTP, not the full stuff.

> therefore it's enough to link maxPacketSize to the configuration.

I think so too.

**Comment by aler9** [comment]
moved here #1601
